Mathlib/Dynamics/PeriodicPts.lean

```diff
@@ -628,42 +630,109 @@ namespace MulAction
+such that `(n • g) +ᵥ a = a`, or `0` if such an `n` does not exist."]
+noncomputable def period (m : M) (a : α) : ℕ := minimalPeriod (fun x => m • x) a
+
+/-- `MulAction.period m a` is definitionally equal to `Function.minimalPeriod (m • ·) a`. -/
+@[to_additive]
+theorem period_eq_minimalPeriod {m : M} {a : α} :
+    MulAction.period m a = minimalPeriod (fun x => m • x) a := rfl
+
+@[to_additive]
+lemma smul_pow_eq_function_pow (m : M) (a : α) (n : ℕ): m ^ n • a = (fun x => m • x)^[n] a := by
+  rw [smul_iterate]
+
+/-- `m ^ (period m a)` fixes `a` -/
+@[to_additive (attr := simp) "`(period m a) • m` fixes `a`"]
+theorem smul_pow_period_fixed (m : M) (a : α) : m ^ (period m a) • a = a := by
+  rw [period_eq_minimalPeriod, smul_pow_eq_function_pow, iterate_minimalPeriod]
+
+@[to_additive]
+lemma fixed_iff_isPeriodicPt {m : M} {a : α} {n : ℕ} :
+    m ^ n • a = a ↔ IsPeriodicPt (fun x => m • x) n a := by
+  rw [smul_pow_eq_function_pow]
+  rfl
+
+/-! ### Multiples of `MulAction.period`
+
+It is easy to convince onself that if `g ^ n • a = a` (resp. `(n • g) +ᵥ a = a`),
+then `n` must be a multiple of `period g a`.
+
+This also holds for negative powers/multiples.
+-/
+
+@[to_additive]
+theorem pow_smul_eq_iff_period_dvd {n : ℕ} {m : M} {a : α}:
+    m ^ n • a = a ↔ period m a ∣ n := by
+  rw [period_eq_minimalPeriod, ← isPeriodicPt_iff_minimalPeriod_dvd, fixed_iff_isPeriodicPt]
+
+@[to_additive]
+theorem zpow_smul_eq_iff_period_dvd {j : ℤ} {g : G} {a : α}:
+    g ^ j • a = a ↔ (period g a : ℤ) ∣ j := by
+  rcases j with n | n
+  · rw [Int.ofNat_eq_coe, zpow_ofNat, Int.coe_nat_dvd, pow_smul_eq_iff_period_dvd]
+  · rw [Int.negSucc_coe, zpow_neg, zpow_ofNat, inv_smul_eq_iff, eq_comm, dvd_neg, Int.coe_nat_dvd,
+      pow_smul_eq_iff_period_dvd]
+
+@[to_additive (attr := simp)]
+theorem pow_smul_plus_period (n o : ℕ) (m : M) (a : α):
+    m ^ (n + (period m a) * o) • a = m ^ n • a := by
+  rw [pow_add, mul_smul, pow_smul_eq_iff_period_dvd.mpr (dvd_mul_right _ _)]
+
+@[to_additive (attr := simp)]
+theorem zpow_smul_plus_period (i j : ℤ) (g : G) (a : α):
+    g ^ (i + (period g a : ℤ) * j) • a = g ^ i • a := by
+  rw [zpow_add, mul_smul, zpow_smul_eq_iff_period_dvd.mpr (dvd_mul_right _ _)]
+
+@[to_additive (attr := simp)]
+theorem pow_smul_mod_period (n : ℕ) {m : M} {a : α}:
+    m ^ (n % period m a) • a = m ^ n • a := by
+  conv_rhs => rw [← Nat.mod_add_div n (period m a), pow_smul_plus_period]
+
+@[to_additive (attr := simp)]
+theorem zpow_smul_mod_period (j : ℤ) {g : G} {a : α}:
+    g ^ (j % (period g a : ℤ)) • a = g ^ j • a := by
+  conv_rhs => rw [← Int.emod_add_ediv j (period g a), zpow_smul_plus_period]
+
+variable {a : G} {b : α}
 
 @[to_additive]
 theorem pow_smul_eq_iff_minimalPeriod_dvd {n : ℕ} :
-    a ^ n • b = b ↔ Function.minimalPeriod (a • ·) b ∣ n := by
-  rw [← isPeriodicPt_iff_minimalPeriod_dvd, IsPeriodicPt, IsFixedPt, smul_iterate]
+    a ^ n • b = b ↔ minimalPeriod (a • ·) b ∣ n := by
```

Comment:
This is what `pow_smul_eq_iff_period_dvd` does above, these theorems are left here for compatibility purposes, since they're used in defining quotient groups, and will be made deprecated in #9514.

Reply:
Ah, great.